adb/adb_commands.py

```diff
@@ -171,19 +172,39 @@ def Pull(self, device_filename, dest_file=None, timeout_ms=None):
     Returns:
       The file data if dest_file is not set.
     """
-    if isinstance(dest_file, basestring):
-      dest_file = open(dest_file, 'w')
-    elif not dest_file:
-      dest_file = cStringIO.StringIO()
-    connection = self.protocol_handler.Open(
-        self.handle, destination='sync:',
-        timeout_ms=timeout_ms)
-    self.filesync_handler.Pull(connection, device_filename, dest_file)
-    connection.Close()
-    # An empty call to cStringIO.StringIO returns an instance of
-    # cStringIO.OutputType.
-    if isinstance(dest_file, cStringIO.OutputType):
-      return dest_file.getvalue()
+
+    filemode, _, _= self.Stat(device_filename)
+
+    if stat.S_ISDIR(filemode):
+      if dest_file is None:
+        raise ValueError("Must specify dest_file when pulling a directory")
+      file_list = self.List(device_filename)
+      if not os.path.exists(dest_file):
+        os.makedirs(dest_file)
+
+      for device_file in file_list:
+        if device_file.filename not in (".", ".."):
+          self.Pull(
+            os.path.join(device_filename, device_file.filename),
+            os.path.join(dest_file, device_file.filename) if dest_file else None
```

Comment:
dest_file can't be None, so are you expecting it to be an empty string? If so, os.path.join('', filename) == filename, so it shouldn't matter. The only other case is if dest_file is False or something falsey but not a string, which should be caught in the check above (by it failing or os.path.exists failing)

Reply:
So from where I'm sitting, this really only means that the conditional at L189 should be removed. Is that correct?